### PR TITLE
Morello

### PR DESCRIPTION
**Pull request by zxombie**
Initial set of Morello changes.

The MORELLO option will be replaced with CPU_CHERI when we build the entire kernel as hybrid.

### CODE REVIEW

File: sys/arm64/include/cherireg.h [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+/*
+ * Root "object-type" capability for the kernel.  This can be used neither as
+ * a data nor code capability.
+ */
+#define	CHERI_PERM_KERN_TYPE	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL |	\
+	CHERI_PERM_UNSEAL)
+
+/*
+ * Basic userspace permission mask; CHERI_PERM_EXECUTE will be added for
+* executable capabilities (pcc); CHERI_PERM_STORE, CHERI_PERM_STORE_CAP,
```
Comment (jrtc27) [inline sys/arm64/include/cherireg.h:95]:
```suggestion
 * executable capabilities (pcc); CHERI_PERM_STORE, CHERI_PERM_STORE_CAP,
```

File: sys/arm64/arm64/exception-morello.S [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	ldp	c8,  c9,  [sp, #(TF_X + 8  * 16)]
+	ldp	c10, c11, [sp, #(TF_X + 10 * 16)]
+	ldp	c12, c13, [sp, #(TF_X + 12 * 16)]
+	ldp	c14, c15, [sp, #(TF_X + 14 * 16)]
+	ldp	c16, c17, [sp, #(TF_X + 16 * 16)]
+.if \el == 0
+	/*
+	 * We only restore the callee saved registers when returning to
+	 * userland as they may have been updated by a system call or signal.
+	 */
```
Comment (jrtc27) [inline sys/arm64/arm64/exception-morello.S:127]: This decision will come back to bite any attempts at kernel compartmentalisation...

File: sys/arm64/arm64/exception-morello.S [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	ldp	c18, c19, [sp, #(TF_X + 18 * 16)]
+	ldp	c20, c21, [sp, #(TF_X + 20 * 16)]
+	ldp	c22, c23, [sp, #(TF_X + 22 * 16)]
+	ldp	c24, c25, [sp, #(TF_X + 24 * 16)]
+	ldp	c26, c27, [sp, #(TF_X + 26 * 16)]
+	ldp	c28, c29, [sp, #(TF_X + 28 * 16)]
+.else
+	ldr	     x29, [sp, #(TF_X + 29 * 16)]
```
Comment (jrtc27) [inline sys/arm64/arm64/exception-morello.S:135]:
```suggestion
	ldr	     c29, [sp, #(TF_X + 29 * 16)]
```

File: sys/arm64/arm64/locore.S [hunk 4]
```diff
@@ -597,10 +618,8 @@ build_l1_block_pagetable:
 
 	/* Build the L1 block entry */
 	orr	x12, x7, #L1_BLOCK
-	orr	x12, x12, #(ATTR_AF)
-#ifdef SMP
-	orr	x12, x12, ATTR_SH(ATTR_SH_IS)
-#endif
+	ldr	x13, =ATTR_DEFAULT
```
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:621]: Has this (and similar diffs) been upstreamed?

File: sys/arm64/arm64/machdep.c [hunk 5]
```diff
@@ -454,16 +470,21 @@ exec_setregs(struct thread *td, struct image_params *imgp, uintcap_t stack)
 	tf->tf_elr = imgp->entry_addr;
 }
 
+#ifndef MORELLO
 /* Sanity check these are the same size, they will be memcpy'd to and fro */
```
Comment (jrtc27) [inline sys/arm64/arm64/machdep.c:474]: Uh, are these assertions not rather important as them failing means there aren't capabilities in the types?

File: sys/arm64/arm64/locore.S [hunk 6]
```diff
@@ -266,7 +277,12 @@ drop_to_el1:
 
 	/* Hypervisor trap functions */
 	adr	x2, hyp_vectors
+#ifdef MORELLO
+	cvtp	c2, x2
```
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:281]:
```suggestion
	/* TODO: Remove when the kernel understands capabilities */
	cvtp	c2, x2
```
?

File: sys/arm64/arm64/locore.S [hunk 7]
```diff
@@ -286,7 +302,12 @@ drop_to_el1:
 2:
 
 	/* Set the address to return to our return address */
+#ifdef MORELLO
+	cvtp	c30, x30
```
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:306]:
```suggestion
	/* TODO: Remove when the kernel understands capabilities */
	cvtp	c30, x30
```
?

File: sys/arm64/arm64/locore.S [hunk 8]
```diff
@@ -665,7 +682,12 @@ start_mmu:
 
 	/* Load the exception vectors */
 	ldr	x2, =exception_vectors
+#ifdef MORELLO
+	cvtp	c2, x2
```
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:686]:
```suggestion
	/* TODO: Remove when the kernel understands capabilities */
	cvtp	c2, x2
```
?

File: sys/arm64/arm64/swtch.S [hunk 9]
```diff
@@ -236,18 +253,33 @@ ENTRY(fork_trampoline)
...
+#else
 	ldp	x0, x1, [sp, #TF_SP]
+#endif
 	msr	sp_el0, x0
 	mov	lr, x1
 
 	/* Restore elr and spsr */
+#ifdef MORELLO
+	ldp	c0, c1, [sp, #TF_ELR]
+	cvtp	c0, x0
```
Comment (jrtc27) [inline sys/arm64/arm64/swtch.S:267]:
```suggestion
	/* TODO: Remove when the kernel understands capabilities */
	cvtp	c0, x0
```
?

File: sys/arm64/include/armreg.h [hunk 10]
```diff
@@ -736,62 +746,109 @@
...
+#define	TCR_EPD0		(1UL << TCR_EPD1_SHIFT)
+/* Bit 6 is reserved */
+#define	TCR_T0SZ_SHIFT		0
+#define	TCR_T0SZ_MASK		0x3f
+#define	TCR_T0SZ(x)		((x) << TCR_T0SZ_SHIFT)
+#define	TCR_TxSZ(x)		(TCR_T1SZ(x) | TCR_T0SZ(x))
 
 #define	TCR_CACHE_ATTRS	((TCR_IRGN0_WBWA | TCR_IRGN1_WBWA) |\
 				(TCR_ORGN0_WBWA | TCR_ORGN1_WBWA))
-
```
Comment (jrtc27) [inline sys/arm64/include/armreg.h:781]: Unnecessary diff

File: sys/arm64/include/cherireg.h [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * $FreeBSD$
+ */
+#ifndef _ARM64_INCLUDE_CHERIREG_H_
```
Comment (jrtc27) [inline sys/arm64/include/cherireg.h:36]: Convention is `_MACHINE_INCLUDE_CHERIREG_H_`. I don't know why ifunc.h and efi.h decide to do `__${ARCH}_INCLUDE_FOO_H` instead but they're the unusual ones, and even if you wanted to follow that you'd need an extra underscore at the start.

File: sys/arm64/include/cherireg.h [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+/*
+ * When performing a userspace-to-userspace CCall, capability flow-control
+ * checks normally prevent local capabilities from being delegated.  This can
+ * be disabled on call (but not return) by using an object type with the 22nd
+ * bit set -- combined with a suitable selector on the CCall instruction to
+ * ensure that this behaviour is intended.
+ */
+#define	CHERI_OTYPE_LOCALOK_SHIFT	(CHERI_OTYPE_BITS - 2)
+#define	CHERI_OTYPE_LOCALOK_FLAG	(1 << CHERI_OTYPE_LOCALOK_SHIFT
+#define	CHERI_OTYPE_IS_LOCALOK(x)	(((x) & CHERI_OTYPE_LOCALOK_FLAG) != 0)
```
Comment (jrtc27) [inline sys/arm64/include/cherireg.h:165]: This is all MIPS-specific (well, the comment at least) and I assume dates back to the old exception-based CCall/CReturn mechanism. RISC-V doesn't have any of this.

File: sys/arm64/arm64/machdep.c [hunk 13]
```diff
@@ -226,7 +237,12 @@ set_regs(struct thread *td, struct reg *regs)
 	frame->tf_spsr &= ~PSR_FLAGS;
 	frame->tf_spsr |= regs->spsr & PSR_FLAGS;
 
+#ifdef MORELLO
+	for (i = 0; i < nitems(frame->tf_x); i++)
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:241]:
```suggestion
	for (int i = 0; i < nitems(frame->tf_x); i++)
```
For loop initializers are fine now (https://reviews.freebsd.org/D25312)

File: sys/arm64/arm64/machdep.c [hunk 14]
```diff
@@ -218,6 +226,9 @@ int
 set_regs(struct thread *td, struct reg *regs)
 {
 	struct trapframe *frame;
+#ifdef MORELLO
+	int i;
+#endif
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:231]:
```suggestion
```

File: sys/arm64/arm64/machdep.c [hunk 15]
```diff
@@ -454,16 +470,21 @@ exec_setregs(struct thread *td, struct image_params *imgp, uintcap_t stack)
...
     sizeof((struct reg *)0)->x);
+#endif
 
 int
 get_mcontext(struct thread *td, mcontext_t *mcp, int clear_ret)
 {
 	struct trapframe *tf = td->td_frame;
+#ifdef MORELLO
+	int i;
+#endif
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:487]:
```suggestion
```

File: sys/arm64/arm64/machdep.c [hunk 16]
```diff
@@ -473,8 +494,13 @@ get_mcontext(struct thread *td, mcontext_t *mcp, int clear_ret)
 		mcp->mc_gpregs.gp_spsr = tf->tf_spsr;
 	}
 
+#ifdef MORELLO
+	for (i = 1; i < nitems(tf->tf_x); i++)
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:498]:
```suggestion
	for (int i = 1; i < nitems(tf->tf_x); i++)
```

File: sys/arm64/arm64/machdep.c [hunk 17]
```diff
@@ -489,14 +515,22 @@ set_mcontext(struct thread *td, mcontext_t *mcp)
 {
 	struct trapframe *tf = td->td_frame;
 	uint32_t spsr;
+#ifdef MORELLO
+	int i;
+#endif
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:520]:
```suggestion
```

File: sys/arm64/arm64/machdep.c [hunk 18]
```diff
@@ -489,14 +515,22 @@ set_mcontext(struct thread *td, mcontext_t *mcp)
...
 
 	spsr = mcp->mc_gpregs.gp_spsr;
 	if ((spsr & PSR_M_MASK) != PSR_M_EL0t ||
 	    (spsr & PSR_AARCH32) != 0 ||
 	    (spsr & PSR_DAIF) != (td->td_frame->tf_spsr & PSR_DAIF))
 		return (EINVAL); 
 
+#ifdef MORELLO
+	for (i = 0; i < nitems(tf->tf_x); i++)
```
Comment (arichardson) [inline sys/arm64/arm64/machdep.c:529]:
```suggestion
	for (int i = 0; i < nitems(tf->tf_x); i++)
```

File: sys/arm64/conf/GENERIC-MORELLO [hunk 19]
```diff
@@ -0,0 +1,227 @@
+#
+# GENERIC -- Generic kernel configuration file for FreeBSD/arm64
```
Comment (arichardson) [inline sys/arm64/conf/GENERIC-MORELLO:2]: Can this include GENERIC and override the differing options?

**Review by jrtc27 — CHANGES_REQUESTED** [review]
How far off is a hybrid kernel? I can't help but feel the use of `__uint128_t` in place of capabilities and not having things like a CHERI-aware memcpy is extremely fragile and just asking for trouble. Personally I'd rather see the full hybrid kernel patch set land as a single pull request, as RISC-V did, rather than committing something with workarounds that are going to be removed/changed very soon, and up until then have the hacky version live on a branch.

File: sys/arm64/include/cherireg.h [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+/*
+ * The CHERI object-type space is split between userspace and kernel,
+ * permitting kernel object references to be delegated to userspace (if
+ * desired).  Currently, we provide 13 bits of namespace to each, with the top
+ * bit set for kernel object types, but it is easy to imagine other splits.
+ * User and kernel software should be written so as to not place assumptions
+ * about the specific values used here, as they may change.
+ */
+#define	CHERI_OTYPE_BITS	(14)
+#define	CHERI_OTYPE_USER_MIN	(0)
```
Comment (arichardson) [inline sys/arm64/include/cherireg.h:148]: This should be 4, since the first 3 are reserved.
```suggestion
#define	CHERI_OTYPE_USER_MIN	(4)
```

**Review by arichardson — CHANGES_REQUESTED** [review]

File: sys/arm64/conf/GENERIC-MORELLO (hunk 19 above)
Comment (zxombie) [reply sys/arm64/conf/GENERIC-MORELLO:2]: We need to remove quite a few drivers that don't build on Morello and will never be used, e.g. any SoC specific driver.

File: sys/arm64/arm64/machdep.c (hunk 13 above)
Comment (zxombie) [reply sys/arm64/arm64/machdep.c:241]: I have future changes that switch may of these back to memcpy. We can clean up then if needed.

File: sys/arm64/include/armreg.h (hunk 10 above)
Comment (zxombie) [reply sys/arm64/include/armreg.h:781]: This is an upstream change I've cherry-picked.

File: sys/arm64/include/cherireg.h (hunk 11 above)
Comment (zxombie) [reply sys/arm64/include/cherireg.h:36]: It's a bit random as which is used.

File: sys/arm64/arm64/locore.S (hunk 8 above)
Comment (zxombie) [reply sys/arm64/arm64/locore.S:686]: All uses of cvtp have an implicit TODO to review later.

File: sys/arm64/arm64/machdep.c (hunk 5 above)
Comment (zxombie) [reply sys/arm64/arm64/machdep.c:474]: They will be fixed by a later change. They are to check the memcpy below them is valid. As there is no memcpy in the Morello case they can be ignored.

File: sys/arm64/conf/GENERIC-MORELLO (hunk 19 above)
Comment (arichardson) [reply sys/arm64/conf/GENERIC-MORELLO:2]: I don't like duplicating all the options lines since we miss all changes to GENERIC. For example we are now missing my "enable TMPFS" change.

Can we at least factor out the options part of the config file into a std.foo and keep the devices list separate?

File: sys/arm64/include/cherireg.h (hunk 11 above)
Comment (jrtc27) [reply sys/arm64/include/cherireg.h:36]: I was just going on
```
jrtc4@zeno:~/cheri/cheribsd$ grep -l _ARM64 sys/arm64/include/*.h
sys/arm64/include/efi.h
sys/arm64/include/ifunc.h
```
ie suggesting we stick to the predominant style there.

**Comment by zxombie** [comment]
Hybrid kernel is close, however I'm still working through extracting the changes needed for it.

**Comment by zxombie** [comment]
Close as #685 should be a better start
